### PR TITLE
[WIP] Overriding reducer of SpatialImage

### PR DESCRIPTION
**Pull request by arthurmensch**
This should sort #1362, but may make things slower. I took to use monkey patch instead of inheritance as we should be able to handle Nifti1Image, Nifti1Pair and what not, so I think this is the most elegant way. `test_image.py` now fails on master.

### CODE REVIEW

File: nilearn/image/image.py [hunk 1]
```diff
@@ -7,23 +7,54 @@
...
+                              affine=state['affine'],
+                              header=state['header'],
+                              extra=state['extra'],
+                              )
+    self.__dict__ = new_self.__dict__
+    if state['filename'] is not None:
+        self.set_filename(state['filename'])
+
+
+def _cachable_niimg_factory(niimg):
```
Comment (AlexandreAbraham) [inline nilearn/image/image.py]: Misleading name, this is not a factory

Comment (arthurmensch) [inline nilearn/image/image.py]: Indeed.

**Comment by arthurmensch** [comment]
I agree this is not a good design : this requires to load in memory all of the nifti images when dispatching them with joblib. But this is the only way to have consistent hashing across pickling, if we are not to modify the joblib hashing code itself (which is what I do in my code, to avoid the memory issue).

**Comment by kchawla-pi** [comment]
Is this alive? If not, please close it by this weekend.

**Comment by tsalo** [comment]
It seems like this is not alive. Closing.

**Pull request closed without merging**